src/Maestro/maestro-angular/src/app/page/build-graph-table/build-graph-table.component.ts

```diff
@@ -96,11 +99,47 @@ function sortBuilds(graph: BuildGraph): BuildData[] {
   result = result.reverse();
   for (const node of result) {
     node.isToolset = isToolset(node.build, graph);
+    node.hasIncoherentDependencies = hasIncoherentDependencies(node.build, result, false);
+    node.hasIncoherentDependenciesIncludingToolsets = hasIncoherentDependencies(node.build, result, true);
   }
 
   return result;
 }
 
+function hasIncoherentDependencies(build: Build, buildData: BuildData[], includeToolsets: boolean): boolean {
+  let y = buildData.find(x => x.build.id == build.id);
+  if(y!.hasIncoherentDependencies)
+  {
+    return true;
+  }
+
+  if (y!.build.dependencies) {
+    for (const dep of y!.build.dependencies) {
+      if(dep && dep!.isProduct){
+        let z = buildData.find(r => r.build.id == dep.buildId);
+
+        if((((!z!.coherent.withProduct && !includeToolsets) || (!z!.coherent.withAll && includeToolsets)) || 
+          ((z!.hasIncoherentDependencies && !includeToolsets) || (z!.hasIncoherentDependenciesIncludingToolsets && includeToolsets))) && 
+          z!.build.dependencies!.length == 0)
+        {
+          return true;
+        }
```

Comment:
This needs to be refactored somehow.  It's a 4 level deep paranthetical if.... I have no idea what it's trying to say, so I can't really recommend a more readable solution.

Reply:
Refactored this to be a bit more readable :)